subsys/usb/class/bt_h4.c

```diff
@@ -0,0 +1,315 @@
+	struct usb_ep_descriptor if0_out_ep;
+	struct usb_ep_descriptor if0_in_ep;
+} __packed;
+
+USBD_CLASS_DESCR_DEFINE(primary, 0) struct usb_bt_h4_config bt_h4_cfg = {
+	/* Interface descriptor 0 */
+	.if0 = {
+		.bLength = sizeof(struct usb_if_descriptor),
+		.bDescriptorType = USB_INTERFACE_DESC,
+		.bInterfaceNumber = 0,
+		.bAlternateSetting = 0,
+		.bNumEndpoints = 2,
+		.bInterfaceClass = CUSTOM_CLASS, /* TBD */
+		.bInterfaceSubClass = 0,
+		.bInterfaceProtocol = 0,
+		.iInterface = 0,
+	},
+
+	/* Data Endpoint OUT */
+	.if0_out_ep = {
+		.bLength = sizeof(struct usb_ep_descriptor),
+		.bDescriptorType = USB_ENDPOINT_DESC,
+		.bEndpointAddress = BT_H4_OUT_EP_ADDR,
+		.bmAttributes = USB_DC_EP_BULK,
+		.wMaxPacketSize = sys_cpu_to_le16(BT_H4_BULK_EP_MPS),
+		.bInterval = 0x01,
+	},
+
+	/* Data Endpoint IN */
+	.if0_in_ep = {
+		.bLength = sizeof(struct usb_ep_descriptor),
+		.bDescriptorType = USB_ENDPOINT_DESC,
+		.bEndpointAddress = BT_H4_IN_EP_ADDR,
+		.bmAttributes = USB_DC_EP_BULK,
+		.wMaxPacketSize = sys_cpu_to_le16(BT_H4_BULK_EP_MPS),
+		.bInterval = 0x01,
+	},
+};
+
+#define H4_CMD 0x01
+#define H4_ACL 0x02
+#define H4_SCO 0x03
+#define H4_EVT 0x04
+
+static struct usb_ep_cfg_data bt_h4_ep_data[] = {
+	{
+		.ep_cb = usb_transfer_ep_callback,
+		.ep_addr = BT_H4_OUT_EP_ADDR,
+	},
+	{
+		.ep_cb = usb_transfer_ep_callback,
+		.ep_addr = BT_H4_IN_EP_ADDR,
+	},
+};
+
+static void bt_h4_read(u8_t ep, int size, void *priv)
+{
+	struct net_buf *buf = priv;
+
+	if (size > 0) {
+		buf->len += size;
+		net_buf_put(&rx_queue, buf);
+		buf = NULL;
+	}
+
+	if (buf) {
+		net_buf_unref(buf);
+	}
+
+	buf = net_buf_alloc(&rx_pool, K_FOREVER);
+
+	net_buf_reserve(buf, BT_BUF_RESERVE);
+
+	/* Start a new read transfer */
+	usb_transfer(bt_h4_ep_data[BT_H4_OUT_EP_IDX].ep_addr, buf->data,
+		     BUF_SIZE, USB_TRANS_READ, bt_h4_read, buf);
+}
+
+static void usb_h4_send(struct net_buf *buf)
+{
+	LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_ACL_IN:
+		net_buf_push_u8(buf, H4_ACL);
+		break;
+	case BT_BUF_EVT:
+		net_buf_push_u8(buf, H4_EVT);
+		break;
+	default:
+		LOG_ERR("Unknown type %u", bt_buf_get_type(buf));
+		net_buf_unref(buf);
+		return;
+	}
+
+	usb_transfer_sync(bt_h4_ep_data[BT_H4_IN_EP_IDX].ep_addr,
+			  buf->data, buf->len, USB_TRANS_WRITE);
+
+	net_buf_unref(buf);
+}
+
+static void hci_tx_thread(void)
+{
+	LOG_DBG("Start USB Bluetooth thread");
+
+	while (true) {
+		struct net_buf *buf;
+
+		buf = net_buf_get(&tx_queue, K_FOREVER);
+		usb_h4_send(buf);
+	}
+}
+
+static void usb_h4_recv(struct net_buf *buf)
+{
+	u8_t type;
+
+	type = net_buf_pull_u8(buf);
+
+	switch (type) {
+	case H4_CMD:
+		bt_buf_set_type(buf, BT_BUF_CMD);
+		break;
+	case H4_ACL:
+		bt_buf_set_type(buf, BT_BUF_ACL_OUT);
+		break;
+	default:
+		LOG_ERR("Unknown H4 type %u", type);
+		return;
+	}
+
+	LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+
```

Comment:
Does the read USB transport modes need to be handled for this driver?
For a USB host that support both it could read the supported command, and if it sees only h4 is supported in knows it doesn't need to change the mode.

Reply:
This does not use the standard Bluetooth class, so the driver would need to know it is already in H:4 mode, in fact even the Read USB Transport would fail if is sent without the header so the driver must know it only works with H:4 mode so that command is not very useful here.